db/tidb/bank_service_quality.go

```diff
@@ -0,0 +1,209 @@
+	var nemesisRecord core.NemesisGeneratorRecord
+	var duration = time.Minute
+	var failCount float64
+	var failCountPerMinuteBeforeInject float64
+
+	var injectionTime time.Time
+	//var recoveryTime time.Time
+
+	var p80 *time.Duration
+	var p90 *time.Duration
+	var p99 *time.Duration
+
+	startTime := ops[0].Time
+	idx := 0
+	for idx < len(ops) {
+		op := ops[idx]
+		if op.Action == core.ReturnOperation {
+			response := op.Data.(bankResponse)
+			if response.Unknown {
+				failCount++
+			}
+		} else if op.Action == core.InvokeNemesis {
+			injectionTime = op.Time
+			nemesisRecord = op.Data.(core.NemesisGeneratorRecord)
+			break
+		} else if op.Action == core.RecoverNemesis {
+			//recoveryTime = op.Time
+			//break
+		}
+		idx++
+	}
+	failCountPerMinuteBeforeInject = failCount / injectionTime.Sub(startTime).Minutes()
+	if idx == len(ops) {
+		log.Panic("expect more events after recovery from nemesis")
+	}
+
+	var window *countWindow
+	metrics := rtoMetrics{nemesis: &nemesisRecord}
+
+	window, metrics.totalMeasureTime = convertToFailSeries(ops[idx:])
+	if window != nil {
+		for true {
+			startTime, failCount, err := window.count(duration)
+			dur := startTime.Sub(injectionTime)
+			if err != nil {
+				if p80 == nil {
+					p80 = &dur
+				}
+				if p90 == nil {
+					p90 = &dur
+				}
+				if p99 == nil {
+					p99 = &dur
+				}
+				break
+			}
+			failCountPerMinute := float64(failCount) / duration.Minutes()
+
+			if 0.8*failCountPerMinute <= failCountPerMinuteBeforeInject && p80 == nil {
+				p80 = &dur
+			}
+			if 0.9*failCountPerMinute <= failCountPerMinuteBeforeInject && p90 == nil {
+				p90 = &dur
+			}
+			if 0.99*failCountPerMinute <= failCountPerMinuteBeforeInject && p99 == nil {
+				p99 = &dur
+				break
+			}
+		}
+		if p80 != nil {
+			metrics.p80 = *p80
+		}
+		if p90 != nil {
+			metrics.p90 = *p90
+		}
+		if p99 != nil {
+			metrics.p99 = *p99
+		}
+	}
+	if err := c.record(&metrics); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (c bankServiceQualityChecker) record(rto *rtoMetrics) error {
+	os.MkdirAll(path.Dir(""), 0755)
+	f, err := os.OpenFile(c.profFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	data, err := json.Marshal(rto.nemesis)
+	if err != nil {
+		return err
+	}
+	if _, err := f.WriteString(fmt.Sprintf("workload:bank,nemesis:%s,measure_time:%.2f,fail_p80:%.2f,fail_p90:%.2f,fail_p99:%.2f,detail:%s\n",
+		rto.nemesis.Name, rto.totalMeasureTime.Seconds(), rto.p80.Seconds(), rto.p90.Seconds(), rto.p99.Seconds(), string(data))); err != nil {
+		return err
+	}
+	return nil
+}
+
+func convertToFailSeries(ops []core.Operation) (*countWindow, time.Duration) {
+	var series []time.Time
+	var startTime time.Time
+	var endTime time.Time
+	for idx := range ops {
+		op := ops[idx]
+		if op.Action == core.ReturnOperation {
+			response, ok := op.Data.(bankResponse)
+			if !ok {
+				continue
+			}
+			if response.Unknown {
```

Comment:
Where does unknown type response come from? Why not drop it?

Reply:
Unknown type response usually means timeout or failure, I treat them as failure response here.